server/aesdsocket.c

```diff
@@ -0,0 +1,204 @@
+    sa.sa_handler = signal_handler;
+
+    if (sigaction(SIGTERM, &sa, NULL) != 0 || sigaction(SIGINT, &sa, NULL) != 0)
+    {
+        syslog(LOG_ERR, "Error setting up signal handler");
+        cleanup_and_exit(EXIT_FAILURE);
+    }
+}
+
+void handle_client_connection(int client_fd)
+{
+    FILE *fp = fopen(LOG_FILE_LOC, "a+");
+    char buffer[MAX_BUFFER_SIZE];
+
+    while (1)
+    {
+        ssize_t bytes_recv = recv(client_fd, buffer, sizeof(buffer), 0);
+        if (bytes_recv <= 0)
+       	{
+            break;
+        }
+
+        fwrite(buffer, bytes_recv, 1, fp);
+
+        if (memchr(buffer, '\n', bytes_recv) != NULL)
+       	{
+            break;
+        }
+    }
+
+    fclose(fp);
+
+    fp = fopen(LOG_FILE_LOC, "r");
+    while (!feof(fp))
+    {
+        ssize_t bytes_read = fread(buffer, 1, sizeof(buffer), fp);
+        if (bytes_read <= 0) {
+            break;
+        }
+
+        send(client_fd, buffer, bytes_read, 0);
+    }
+
+    fclose(fp);
+    close(client_fd);
+}
+
+void daemonize()
+{
+    pid_t pid = fork();
+
+    if (pid < 0)
+    {
+        syslog(LOG_ERR, "Error forking: %m");
+        cleanup_and_exit(EXIT_FAILURE);
+    } else if (pid > 0)
+    {
+        exit(EXIT_SUCCESS); // Parent exits
+    }
+
+    if (setsid() == -1)
+    {
+        syslog(LOG_ERR, "Error creating new session: %m");
+        cleanup_and_exit(EXIT_FAILURE);
+    }
+
+    chdir("/"); // Change working directory to root
+
+    // Close standard file descriptors
+    close(STDIN_FILENO);
+    close(STDOUT_FILENO);
+    close(STDERR_FILENO);
+
+    // Redirect stdin/out/err to /dev/null
+    int dev_null = open("/dev/null", O_RDWR);
+    if (dev_null == -1)
+    {
+        syslog(LOG_ERR, "Error opening /dev/null: %m");
+        cleanup_and_exit(EXIT_FAILURE);
+    }
+
+    dup2(dev_null, STDIN_FILENO);
+    dup2(dev_null, STDOUT_FILENO);
+    dup2(dev_null, STDERR_FILENO);
+
+    close(dev_null);
+}
+
+int main(int argc, char* argv[])
+{
+    openlog("AESD Socket", LOG_PID | LOG_NDELAY, LOG_USER);
+
+    bool daemon = false;
+
+    if (argc > 1 && strcmp(argv[1], "-d") == 0)
+    {
+        daemon = true;
+    }
+
+    setup_signal_handler();
+
+    struct sockaddr_in server_addr;
+    memset(&server_addr, 0, sizeof(server_addr));
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_port = htons(atoi(PORT));
+    server_addr.sin_addr.s_addr = INADDR_ANY;
+
+    sockfd = socket(AF_INET, SOCK_STREAM, 0);
+    if (sockfd == -1)
+    {
+        syslog(LOG_ERR, "Error creating socket: %m");
+        cleanup_and_exit(EXIT_FAILURE);
+    }
+
+    int optval = 1;
+    if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)) != 0)
+    {
+        syslog(LOG_ERR, "Error setting socket options: %m");
+        cleanup_and_exit(EXIT_FAILURE);
+    }
+
+    if (daemon)
+    {
+        daemonize();
+    }
+
+    if (bind(sockfd, (struct sockaddr*)&server_addr, sizeof(server_addr)) != 0)
```

Comment:
@akka2103 with bind after daemonize() what will happen if we can't bind to the socket?